test_clear_logs is flaky

<!-- READ THIS FIRST:
  - If you need additional help with this template, please refer to https://www.home-assistant.io/help/reporting_issues/
  - Make sure you are running the latest version of Home Assistant before reporting an issue: https://github.com/home-assistant/core/releases
  - Do not report issues for integrations if you are using custom components or integrations.
  - Provide as many details as possible. Paste logs, configuration samples and code into the backticks.
  DO NOT DELETE ANY TEXT from this template! Otherwise, your issue may be closed without comment.
-->
## The problem
<!-- 
  Describe the issue you are experiencing here to communicate to the
  maintainers. Tell us what you were trying to do and what happened.
-->

The `test_clear_logs` test is flaky. Failed once, didn't fail on the second try.

## Environment
<!--
  Provide details about the versions you are using, which helps us to reproduce
  and find the issue quicker. Version information is found in the
  Home Assistant frontend: Configuration -> Info.
-->

- Home Assistant Core release with the issue: 0.114.4
- Last working Home Assistant Core release (if known): no idea
- Operating environment (OS/Container/Supervised/Core): NixOS 20.09
- Integration causing this issue:  n/a
- Link to integration documentation on our website: n/a


## Traceback/Error logs
<!--
  If you come across any trace or error logs, please provide them.
-->

```txt
=================================== FAILURES ===================================
___________________________ test_clear_logs[pyloop] ____________________________

hass = <homeassistant.core.HomeAssistant object at 0x7ffff3c1b220>
simple_queue = <_queue.SimpleQueue object at 0x7ffff0651630>
hass_client = <function hass_client.<locals>.auth_client at 0x7ffff17261f0>

    async def test_clear_logs(hass, simple_queue, hass_client):
        """Test that the log can be cleared via a service call."""
        await async_setup_component(hass, system_log.DOMAIN, BASIC_CONFIG)
        _LOGGER.error("error message")
        await _async_block_until_queue_empty(hass, simple_queue)

        hass.async_add_job(
            hass.services.async_call(system_log.DOMAIN, system_log.SERVICE_CLEAR, {})
        )
        await _async_block_until_queue_empty(hass, simple_queue)

        # Assert done by get_error_log
>       await get_error_log(hass, hass_client, 0)

tests/components/system_log/test_init.py:237:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

hass = <homeassistant.core.HomeAssistant object at 0x7ffff3c1b220>
hass_client = <function hass_client.<locals>.auth_client at 0x7ffff17261f0>
expected_count = 0

    async def get_error_log(hass, hass_client, expected_count):
        """Fetch all entries from system_log via the API."""

        client = await hass_client()
        resp = await client.get("/api/error/all")
        assert resp.status == 200

        data = await resp.json()

>       assert len(data) == expected_count
E       AssertionError: assert 1 == 0
E        +  where 1 = len([{'count': 1, 'exception': '', 'first_occurred': 1599952993.326157, 'level': 'ERROR', ...}])

tests/components/system_log/test_init.py:46: AssertionError
---------------------------- Captured stderr setup -----------------------------
DEBUG:asyncio:Using selector: EpollSelector
------------------------------ Captured log setup ------------------------------
DEBUG    asyncio:selector_events.py:59 Using selector: EpollSelector
----------------------------- Captured stderr call -----------------------------
INFO:homeassistant.loader:Loaded test from custom_components.test
INFO:homeassistant.loader:Loaded test_package from custom_components.test_package
INFO:homeassistant.loader:Loaded system_log from homeassistant.components.system_log
INFO:homeassistant.loader:Loaded http from homeassistant.components.http
DEBUG:homeassistant.setup:Dependency system_log will wait for ['http']
INFO:homeassistant.setup:Setting up http
INFO:homeassistant.setup:Setup of domain http took 0.0 seconds
DEBUG:homeassistant.core:Bus:Handling <Event component_loaded[L]: component=http>
INFO:homeassistant.setup:Setting up system_log
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=system_log, service=clear>
DEBUG:homeassistant.core:Bus:Handling <Event service_registered[L]: domain=system_log, service=write>
INFO:homeassistant.setup:Setup of domain system_log took 0.0 seconds
DEBUG:homeassistant.core:Bus:Handling <Event component_loaded[L]: component=system_log>
ERROR:test_logger:error message
DEBUG:homeassistant.core:Bus:Handling <Event call_service[L]: domain=system_log, service=clear, service_data=>
DEBUG:homeassistant.components.http.auth:Authenticated 127.0.0.1 for /api/error/all using bearer token
DEBUG:homeassistant.components.http.view:Serving /api/error/all to 127.0.0.1 (auth: True)
INFO:aiohttp.access:127.0.0.1 [12/Sep/2020:23:23:13 +0000] "GET /api/error/all HTTP/1.1" 200 339 "-" "Python/3.8 aiohttp/3.6.2"
------------------------------ Captured log call -------------------------------
INFO     homeassistant.loader:loader.py:267 Loaded test from custom_components.test
INFO     homeassistant.loader:loader.py:267 Loaded test_package from custom_components.test_package
INFO     homeassistant.loader:loader.py:267 Loaded system_log from homeassistant.components.system_log
INFO     homeassistant.loader:loader.py:267 Loaded http from homeassistant.components.http
DEBUG    homeassistant.setup:setup.py:87 Dependency system_log will wait for ['http']
INFO     homeassistant.setup:setup.py:159 Setting up http
INFO     homeassistant.setup:setup.py:210 Setup of domain http took 0.0 seconds
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event component_loaded[L]: component=http>
INFO     homeassistant.setup:setup.py:159 Setting up system_log
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event service_registered[L]: domain=system_log, service=clear>
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event service_registered[L]: domain=system_log, service=write>
INFO     homeassistant.setup:setup.py:210 Setup of domain system_log took 0.0 seconds
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event component_loaded[L]: component=system_log>
ERROR    test_logger:test_init.py:228 error message
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event call_service[L]: domain=system_log, service=clear, service_data=>
DEBUG    homeassistant.components.http.auth:auth.py:119 Authenticated 127.0.0.1 for /api/error/all using bearer token
DEBUG    homeassistant.components.http.view:view.py:118 Serving /api/error/all to 127.0.0.1 (auth: True)
INFO     aiohttp.access:web_log.py:233 127.0.0.1 [12/Sep/2020:23:23:13 +0000] "GET /api/error/all HTTP/1.1" 200 339 "-" "Python/3.8 aiohttp/3.6.2"
--------------------------- Captured stderr teardown ---------------------------
DEBUG:homeassistant.core:Bus:Handling <Event homeassistant_stop[L]>
DEBUG:homeassistant.core:Bus:Handling <Event homeassistant_final_write[L]>
INFO:tests.common:Writing data to auth: {'version': 1, 'key': 'auth', 'data': {'users': [{'id': '9eb4b0411d02499da2e0f7395e8eba12', 'group_ids': ['system-admin'], 'is_owner': False, 'is_active': True, 'name': 'Mock User', 'system_generated': False}], 'groups': [{'id': 'system-admin', 'name': 'Administrators'}, {'id': 'system-users', 'name': 'Users'}, {'id': 'system-read-only', 'name': 'Read Only'}], 'credentials': [], 'refresh_tokens': [{'id': 'bed5072c9158470ea7caaeb2dbd4d699', 'user_id': '9eb4b0411d02499da2e0f7395e8eba12', 'client_id': 'https://example.com/app', 'client_name': None, 'client_icon': None, 'token_type': 'normal', 'created_at': '2020-09-12T23:23:13.315148+00:00', 'access_token_expiration': 1800.0, 'token': '1bbea525fb267f3b8af85654bda4034e14e24d99dc42df230510a9a51b5cc2483868dee45249e84a770a680cde59c1ed580d0c478e6ccc7501cbed48824444ef', 'jwt_key': '7ac93173388c2ea39274bb2ab39095f46aaabecb9e32366ce9d78f4ab19ffe6b0e5fccae216909aaeea19de8b32ee2a99ee4fa471ef82c786ce526426e89714d', 'last_used_at': '2020-09-12T23:23:13.315275+00:00', 'last_used_ip': None}]}}
DEBUG:homeassistant.core:Bus:Handling <Event homeassistant_close[L]>
---------------------------- Captured log teardown -----------------------------
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event homeassistant_stop[L]>
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event homeassistant_final_write[L]>
INFO     tests.common:common.py:935 Writing data to auth: {'version': 1, 'key': 'auth', 'data': {'users': [{'id': '9eb4b0411d02499da2e0f7395e8eba12', 'group_ids': ['system-admin'], 'is_owner': False, 'is_active': True, 'name': 'Mock User', 'system_generated': False}], 'groups': [{'id': 'system-admin', 'name': 'Administrators'}, {'id': 'system-users', 'name': 'Users'}, {'id': 'system-read-only', 'name': 'Read Only'}], 'credentials': [], 'refresh_tokens': [{'id': 'bed5072c9158470ea7caaeb2dbd4d699', 'user_id': '9eb4b0411d02499da2e0f7395e8eba12', 'client_id': 'https://example.com/app', 'client_name': None, 'client_icon': None, 'token_type': 'normal', 'created_at': '2020-09-12T23:23:13.315148+00:00', 'access_token_expiration': 1800.0, 'token': '1bbea525fb267f3b8af85654bda4034e14e24d99dc42df230510a9a51b5cc2483868dee45249e84a770a680cde59c1ed580d0c478e6ccc7501cbed48824444ef', 'jwt_key': '7ac93173388c2ea39274bb2ab39095f46aaabecb9e32366ce9d78f4ab19ffe6b0e5fccae216909aaeea19de8b32ee2a99ee4fa471ef82c786ce526426e89714d', 'last_used_at': '2020-09-12T23:23:13.315275+00:00', 'last_used_ip': None}]}}
DEBUG    homeassistant.core:core.py:645 Bus:Handling <Event homeassistant_close[L]>
```

## Additional information

--- a/homeassistant/components/system_log/__init__.py
+++ b/homeassistant/components/system_log/__init__.py
@@ -211,6 +211,8 @@ async def async_setup(hass, config):
 
     handler = LogErrorHandler(hass, conf[CONF_MAX_ENTRIES], conf[CONF_FIRE_EVENT])
 
+    hass.data[DOMAIN] = handler
+
     listener = logging.handlers.QueueListener(
         simple_queue, handler, respect_handler_level=True
     )
@@ -222,6 +224,7 @@ async def async_setup(hass, config):
         """Cleanup handler."""
         logging.root.removeHandler(queue_handler)
         listener.stop()
+        del hass.data[DOMAIN]
 
     hass.bus.async_listen_once(EVENT_HOMEASSISTANT_CLOSE, _async_stop_queue_handler)
 

--- a/tests/components/system_log/test_init.py
+++ b/tests/components/system_log/test_init.py
@@ -31,6 +31,8 @@ async def _async_block_until_queue_empty(hass, sq):
     await hass.async_block_till_done()
     while not sq.empty():
         await asyncio.sleep(0.01)
+    hass.data[system_log.DOMAIN].acquire()
+    hass.data[system_log.DOMAIN].release()
     await hass.async_block_till_done()
 
 

